File: 133B/machine.cs
////////////////////////////////////////////////////////////////
using System;
////////////////////////////////////////////////////////////////
class Program
{
    static int getOffset( char c ){
        return c - '`';
    }
    static int getMaxOffset( string f )
    {
        int maxOffset = 0;
        foreach( char c in f )
        {
            int offset = getOffset( c ); 
            if( offset > maxOffset ){
                maxOffset = offset;
            }
        }
        return maxOffset;
    }
    static void Main( string[] args )
    {
        string f = "(acfg)(bcd)(aed)(fade)(bgfae)";
        Console.WriteLine( "formula: " + f );
        int n = getMaxOffset( f );
        Console.WriteLine( "maxOffset: " + n );
        int[] T = new int[ n + 1 ];
        for( int k = 1; k <= n; ++k ){
            T[ k ] = k;
        }
        int j = 0;
        int Z = 0;
        for( int k = f.Length - 1; k >= 0; --k ){
            char c = f[ k ];
            int  i = getOffset( c );
            if( c == ')' ){
                Z = 0;
            } else if( c == '(' ){
                T[ j ] = Z;
            } else {
                ( T[ i ], Z ) = ( Z, T[ i ]);
                if( T[ i ] == 0 ){
                    j = i;
                }
            }
        }
        Console.WriteLine( "[" + string.Join( ", ", T ) + "]" );
    }
}
////////////////////////////////////////////////////////////////
//
//

File: 133B/machine.php
<?php
///////////////////////////////////////////////////////_
$f1  = "(acfg)(bcd)(aed)(fade)(bgfae)";
///////////////////////////////////////////////////////=
// 1-based offset from 'a'
function GetOffset( $c ){
    return ord( $c ) - ord( "a" ) + 1;
}
///////////////////////////////////////////////////////~
// get maximum 1-based offset from 'a' in f
function GetMaxOffset( $f ){
    $m = 0;
    for( $j = strlen( $f ) - 1; $j >= 0; $j-- ){
        $c = $f[ $j ];
        $offset = GetOffset( $c );
        $m = max( $m, $offset );
    }
    return $m;
}
///////////////////////////////////////////////////////^
function ConsTable( $n ){
    for( $j = 0; $j <= $n; $j++ ){
        $a[] = $j;
    }
    return $a;    
}
$n = GetMaxOffset( $f1 );
$j = -1;
$Z = -1;
$T = ConsTable( $n );
for( $k = strlen( $f1 ) - 1; $k >= 0; $k-- ){
    $c = $f1[ $k ];
    if( $c == ")" ){
        $Z = 0;
    } else if( $c == "("){
        $T[ $j ] = $Z;
    } else {
        $i = GetOffset( $c );
        [ $T[ $i ], $Z ] = [ $Z, $T[ $i ]];
        if( $T[ $i ] == 0 ){
            $j = $i;
        }
    }
}
print_r( $T );
///////////////////////////////////////////////////////>
// I think I'm going to drop MySQL:) and replace it with
// some other language:)
// Array ( [0] => 0 [1] => 4 [2] => 3 [3] => 5
//         [4] => 7 [5] => 2 [6] => 6 [7] => 1 )
?>
